### PR TITLE
PR on nnlda layer 

### PR DESCRIPTION
**Pull request by advikavs**
## Description

PR on nnlda layer

### CODE REVIEW

File: deepchem/models/dft/nnxc.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    outmultmode: int
+        The mode to decide Eks from NN output
+    """
+
+    # default value of xcstr is lda_x
+    def __init__(
+            self,
+            xcstr: str,
+            nnmodel: torch.nn.Module,
+            *,
```
Comment (rbharath) [inline deepchem/models/dft/nnxc.py:81]: * arguments are a bad pattern imo. We shouldn't use them in the deepchem codebase.

File: deepchem/models/dft/nnxc.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+            aweight0: float = 0.0,  # weight of the neural network
+            bweight0: float = 1.0,  # weight of the default xc
+            dtype: torch.dtype = torch.double):
+        # hybrid libxc and neural network xc where it starts as libxc and then
+        # trains the weights of libxc and nn xc
+
+        super().__init__()
+        # What is type of xc here?
+        self.xc = get_xc(xcstr)
+        print(type(self.xc))
```
Comment (rbharath) [inline deepchem/models/dft/nnxc.py:95]: Print statements should be removed before merge

File: deepchem/models/dft/nnxc.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        super().__init__()
+        self.nnmodel = nnmodel
+        self.ninpmode = ninpmode
+        self.outmultmode = outmultmode
+
+    @property
+    def family(self) -> int:
+        return 1
+
+    def get_edensityxc(
```
Comment (rbharath) [inline deepchem/models/dft/nnxc.py:37]: Should this method be public or private? Do we expect a user to use this?

**Review by rbharath — COMMENTED** [review]
Good start! Made a couple of preliminary comments

File: deepchem/models/dft/nnxc.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+import torch
+import numpy as np
+from typing import Union, Iterator, List
+from dqc.xc.base_xc import BaseXC
+from dqc.utils.datastruct import ValGrad, SpinParam
+from dqc.utils.safeops import safenorm, safepow
+from dqc.api.getxc import get_xc
+
+
+class NNLDA(BaseXC, torch.nn.Module):
```
[the remark this answers is not in the document]
Comment (advikavs) [reply deepchem/models/dft/nnxc.py:11]: Yes, DQC uses dual inheritance as well.

File: deepchem/models/dft/nnxc.py (hunk 3 above)
Comment (advikavs) [reply deepchem/models/dft/nnxc.py:55]: No, but it is used in HybridXC .

File: deepchem/models/dft/nnxc.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    raise ModuleNotFoundError("This layer requires dqc and torch")
+
+
+class BaseNNXC(BaseXC, torch.nn.Module):
+    """
+    Base class for the NNLDA and HybridXC classes.
+    """
+
+    @abstractproperty
+    def family(self) -> int:
```
Comment (rbharath) [inline deepchem/models/dft/nnxc.py:18]: There should be a docstring here

File: deepchem/models/dft/nnxc.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        pass
+
+    @abstractmethod
+    def get_edensityxc(
```
Comment (rbharath) [inline deepchem/models/dft/nnxc.py:22]: There should be a docstring here

File: deepchem/models/dft/nnxc.py [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+try:
+    from dqc.xc.base_xc import BaseXC
+    from dqc.utils.datastruct import ValGrad, SpinParam
+    from dqc.api.getxc import get_xc
+    import torch
+except ModuleNotFoundError:
+    raise ModuleNotFoundError("This layer requires dqc and torch")
+
+
+class BaseNNXC(BaseXC, torch.nn.Module):
```
Comment (rbharath) [inline deepchem/models/dft/nnxc.py:12]: All these layers should be added to the `docs/` folder so they render in the layer documentation

File: deepchem/models/dft/nnxc.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+    @abstractmethod
+    def get_edensityxc(
+            self, densinfo: Union[ValGrad, SpinParam[ValGrad]]) -> torch.Tensor:
+        pass
+
+
+class NNLDA(BaseNNXC):
+    """
+    Neural network xc functional of LDA (only receives the density as input)
```
Comment (rbharath) [inline deepchem/models/dft/nnxc.py:29]: Can you add a usage example? There should be an explanation of what LDA is as a DFT method, along with a suitable reference. Aim for at least a few paragraphs of discussion

File: deepchem/models/dft/nnxc.py [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+                 ninpmode: int = 1,
+                 outmultmode: int = 1):
+        super().__init__()
+        """
+        Parameters
+        ----------
+        nnmodel: torch.nn.Module
+            Neural network for xc functional
+        ninpmode: int
+            The mode to decide the transformation of the density to NN input.
```
Comment (rbharath) [inline deepchem/models/dft/nnxc.py:43]: What are the valid values? What does each value mean?

File: deepchem/models/dft/nnxc.py [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+        outmultmode: int
```
Comment (rbharath) [inline deepchem/models/dft/nnxc.py:44]: What are the valid values for the output? What does each value mean?

File: deepchem/models/dft/nnxc.py [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+        x = torch.cat((ninp, xi), dim=-1)  # (*BD, nr, 2)
+        nnout = self.nnmodel(x)  # (*BD, nr, 1)
+        res = nnout * n  # (*BD, nr, 1)
+        res = res.squeeze(-1)
+        return res
+
+
+class HybridXC(BaseNNXC):
+    """
+    The HybridXC module computes XC energy by summing XC energy computed
```
Comment (rbharath) [inline deepchem/models/dft/nnxc.py:91]: Can you add a usage example? Can you also extend the explanation of what this layer does?

File: deepchem/models/dft/nnxc.py [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+                 ninpmode: int = 1,
+                 outmultmode: int = 1,
+                 aweight0: float = 0.0,
+                 bweight0: float = 1.0):
+
+        super().__init__()
+        """
+        Parameters
+        ----------
+        xcstr: str
```
Comment (rbharath) [inline deepchem/models/dft/nnxc.py:107]: What are the possible values of `xcstr`?

File: deepchem/models/dft/nnxc.py [hunk 13]
```diff
@@ -0,0 +1,152 @@
+from abc import abstractproperty, abstractmethod
+from typing import Union
+try:
+    from dqc.xc.base_xc import BaseXC
+    from dqc.utils.datastruct import ValGrad, SpinParam
+    from dqc.api.getxc import get_xc
```
Comment (rbharath) [inline deepchem/models/dft/nnxc.py:6]: This looks like a utility function. Can you copy this in directly from dqc?

**Review by rbharath — COMMENTED** [review]
Getting closer! Main issue is we need a lot more documentation to explain what these layers do.

File: deepchem/models/dft/nnxc.py (hunk 10 above)
Comment (advikavs) [reply deepchem/models/dft/nnxc.py:44]: Currently we only support one mode, so I have removed it. Will add it back in the next iterations.

File: deepchem/models/dft/nnxc.py (hunk 13 above)
Comment (advikavs) [reply deepchem/models/dft/nnxc.py:6]: Are we trying to do that in this iteration? If so, I'm not so sure, but I will look into it.

File: deepchem/models/dft/nnxc.py [hunk 14]
```diff
@@ -0,0 +1,199 @@
...
+            The choice of xc to use. The possible values can be found under
+            the "LDA Functionals" section in the reference given below.
+        nnmodel: nn.Module
+            trainable neural network for prediction xc energy.
+        aweight0: float
+            weight of the neural network
+        bweight0: float
+            weight of the default xc
+
+        References
```
Comment (rbharath) [inline deepchem/models/dft/nnxc.py:166]: This should be formatted properly as a reference section

File: deepchem/models/dft/nnxc.py (hunk 12 above)
Comment (rbharath) [reply deepchem/models/dft/nnxc.py:156]: Can you list explicitly the values that are allowed here? Better to not force users to read a second reference which may be complicated

**Review by rbharath — COMMENTED** [review]
Almost there. Just a couple more comment improvements to address and we should be ready to merge

File: deepchem/models/dft/nnxc.py [hunk 15]
```diff
@@ -0,0 +1,183 @@
+from abc import abstractmethod
+from typing import Union
+import torch
+from dqc.utils.datastruct import ValGrad, SpinParam
+from dqc.api.getxc import get_xc
+
+
+class BaseNNXC(torch.nn.Module):
+    """
+    Base class for the NNLDA and HybridXC classes.
```
Comment (rbharath) [inline deepchem/models/dft/nnxc.py:10]: Can you add more detail on when we would want to use this base class and what capabilities it provides? A good paragraph of explanation would be my rule of thumb

File: deepchem/models/dft/nnxc.py [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+        ----------
+        densinfo: Union[ValGrad, SpinParam[ValGrad]]
+            Density information calculated using DQC utilities.
+        """
+        pass
+
+
+class NNLDA(BaseNNXC):
+    """
+    Neural network xc functional of LDA
```
Comment (rbharath) [inline deepchem/models/dft/nnxc.py:30]: Can you add some more basic context? That is, the basic set up that this is a class relevant for density functional theory, a basic explanation of what an exchange correlation functional is?

File: deepchem/models/dft/nnxc.py [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+        the neural network.
+
+        Parameters
+        ----------
+        densinfo: Union[ValGrad, SpinParam[ValGrad]]
+            Density information calculated using DQC utilities.
+
+        Returns
+        -------
+        res
```
Comment (rbharath) [inline deepchem/models/dft/nnxc.py:89]: What is the shape of `res`?

File: deepchem/models/dft/nnxc.py [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+        nnout = self.nnmodel(x)  # (*BD, nr, 1)
+        res = nnout * n  # (*BD, nr, 1)
+        res = res.squeeze(-1)
+        return res
+
+
+class HybridXC(BaseNNXC):
+    """
+    The HybridXC module computes XC energy by summing XC energy computed
+    from libxc and the trainable neural network with tunable weights.
```
Comment (rbharath) [inline deepchem/models/dft/nnxc.py:113]: Can you add a brief explanation of what `libxc` is?

**Review by rbharath — COMMENTED** [review]
Almost ready to go. Have a few more requests for documentation improvement here

**Review by rbharath — COMMENTED** [review]
LGTM, feel free to merge in once CI is clear

**Comment by advikavs** [comment]
> LGTM, feel free to merge in once CI is clear

The CI is fine. But I don't have merge access, it is blocked. Thanks

**Review by rbharath — APPROVED** [review]
LGTM

**Pull request merged**